src/audio/google/google_rtc_audio_processing.c

```diff
@@ -822,23 +826,66 @@ static int google_rtc_audio_processing_process(struct processing_module *mod,
 	/* 32float: de-interlace ref buffer, convert it to float, skip channels if > Max
 	 * 16int: linearize buffer, skip channels if > Max
 	 */
+	/* Reduce cycle waste by streamlining the inner loop,
+	 * converting from array indexing to pointer arithmetic,
+	 * and putting data copy verification outside the loop.
+	 */
 	buffer_offset = 0;
-	for (int i = 0; i < cd->num_frames; i++) {
-		for (channel = 0; channel < cd->num_aec_reference_channels; ++channel) {
+	const int16_t *ref_end = ref + cd->num_frames * cd->num_aec_reference_channels;
+
+	/* Check if the calculated end of the reference buffer exceeds the actual end
+	 * of the buffer
+	 */
+	if (ref_end >= (const int16_t *)ref_buf_end)
+		/* If it does, wrap the reference buffer end pointer back to the start
+		 * of the buffer
+		 */
+		ref_end = (const int16_t *)ref_buf_start;
+
 #if CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API
-			cd->aec_reference_buffer_ptrs[channel][i] =
-					convert_int16_to_float(ref[channel]);
+	float **ref_ptr = cd->aec_reference_buffer_ptrs;
+	int s_chan;
+	int i;
+
+	/* Loop over frames and channels, converting data from int16 to float */
+	for (i = 0; i < cd->num_frames; ++i) {
+		for (s_chan = 0; s_chan < cd->num_aec_reference_channels; ++s_chan) {
+			/* Check that ref is within the valid range of the ref_buf buffer */
+			if ((int16_t const *)ref >= ref_buf_start &&
+			    (int16_t const *)ref < ref_buf_end) {
+				(*ref_ptr)[s_chan] = convert_int16_to_float(*ref++);
+			} else {
+				/* ref does not point to valid int16_t data */
+				return ERR_INVALID_REF;
+			}
+	    ref_ptr++;
+	}
+
 #else /* CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API */
-			cd->aec_reference_buffer[buffer_offset++] = ref[channel];
-#endif /* CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API */
+	int16_t *ref_buf = cd->aec_reference_buffer;
 
-		}
+	/* Check that ref is within the valid range of the ref_buf buffer */
+	if (ref && ref >= (int16_t *)ref_buf_start && ref < (int16_t *)ref_buf_end) {
+		/* Use memcpy_s to copy the data from ref buffer to ref_buf buffer until it reaches
+		 * ref_end
+		 * This assumes that the data in the ref buffer is contiguous
+		 */
+		size_t num_bytes = (ref_end - ref) * sizeof(*ref);
 
-		ref += cd->num_aec_reference_channels;
-		if ((void *)ref >= (void *)ref_buf_end)
-			ref = (void *)ref_buf_start;
+		if (memcpy_s(ref_buf, num_bytes, ref, num_bytes) != 0) {
```

Comment:
I would not use mempcy_s() in hot code parts. There's num_bytes in both source and dest size so memcpy_s() never will return error. This seems to apply for all memcpy_s() usages in this patch.